File: 1496-lucky-numbers-in-a-matrix/1496-lucky-numbers-in-a-matrix.py
class Solution:
    def luckyNumbers(self, matrix):
        n = len(matrix)
        m = len(matrix[0])
        lucky_numbers = []

        for i in range(n):
            num = float('inf')
            index = -1

            # Find the minimum number in the current row
            for j in range(m):
                if matrix[i][j] < num:
                    num = matrix[i][j]
                    index = j

            # Check if this number is the maximum in its column
            flag = True
            for row in range(n):
                if matrix[row][index] > num:
                    flag = False
                    break

            if flag:
                lucky_numbers.append(num)

        return lucky_numbers

File: 1496-lucky-numbers-in-a-matrix/1496-lucky-numbers-in-a-matrix.java
class Solution {
    public List<Integer> luckyNumbers(int[][] matrix) {
        List<Integer> res = new ArrayList<>();

        // Iterate through each row of the matrix
        for (int i = 0; i < matrix.length; i++) {
            int rowMin = matrix[i][0]; // Initialize the minimum element of the row
            int colIndex = 0; // To keep track of the column index of the minimum element

            // Finding the smallest element in the ith row
            for (int j = 1; j < matrix[i].length; j++) {
                if (matrix[i][j] < rowMin) {
                    rowMin = matrix[i][j];
                    colIndex = j; // Update column index
                }
            }

            // Checking if rowMin is also the largest element in its column
            boolean isLuckyNumber = true; // Flag to check if it's a lucky number
            for (int j = 0; j < matrix.length; j++) {
                if (matrix[j][colIndex] > rowMin) {
                    isLuckyNumber = false; // Not a lucky number
                    break;
                }
            }

            // If it's a lucky number, add it to the result list
            if (isLuckyNumber) {
                res.add(rowMin);
            }
        }

        return res; // Return the list of lucky numbers
    }
}